### PR TITLE
Rename wallet data

### LINKED ISSUE #1903: Rename `WalletData`
## Description

Find a better name for the `WalletData` struct.

## Motivation

`WalletData` is rather unspecific. Ideally the new name would better insinuate what it's used for.

## Are you planning to do it yourself in a pull request?

Yes/No.

### PR DESCRIPTION
**Pull request by Alex6323**
# Description of change

1. Renames `WalletData` to `WalletLedger` and corresponding field/variable names.
2. Moves `address`, `bip_path`, and `alias` out of `WalletLedger` into `Wallet`

## Links to any relevant issues

Closes #1903

### CODE REVIEW

File: bindings/core/src/method_handler/call_method.rs [hunk 1]
```diff
@@ -58,7 +58,7 @@ pub async fn call_client_method(client: &Client, method: ClientMethod) -> Respon
 }
 
 /// Call a wallet method.
-pub async fn call_wallet_method(wallet: &Wallet, method: WalletMethod) -> Response {
+pub async fn call_wallet_method(wallet: &mut Wallet, method: WalletMethod) -> Response {
```
Comment (Thoralf-M) [inline bindings/core/src/method_handler/call_method.rs]: Requiring it mutable looks annoying

Comment (Thoralf-M) [reply bindings/core/src/method_handler/call_method.rs]: It's actually pretty bad because it blocks everything in the bindings

Comment (Thoralf-M) [reply bindings/core/src/method_handler/call_method.rs]: Just read the PR description, don't we only need mutability for the alias, why not just add a lock for it?

File: cli/src/cli.rs [hunk 2]
```diff
@@ -511,7 +517,7 @@ pub async fn restore_command_stronghold(
     }
 }
 
-pub async fn set_node_url_command(wallet: &Wallet, url: String) -> Result<(), Error> {
+pub async fn set_node_url_command(wallet: &mut Wallet, url: String) -> Result<(), Error> {
```
Comment (Thoralf-M) [inline cli/src/cli.rs]: `mut` still needed here?

File: bindings/wasm/src/wallet.rs [hunk 3]
```diff
@@ -65,7 +65,7 @@ pub async fn get_secret_manager(method_handler: &WalletMethodHandler) -> Result<
 #[wasm_bindgen(js_name = callWalletMethod)]
 pub async fn call_wallet_method(method_handler: &WalletMethodHandler, method: String) -> Result<String, JsError> {
     let method = serde_json::from_str(&method).map_err(map_err)?;
-    match &*method_handler.0.read().await {
+    match &mut *method_handler.0.write().await {
```
Comment (Thoralf-M) [inline bindings/wasm/src/wallet.rs]: Mutability needed here? Was `match &*method_handler.0.read().await {` before

File: bindings/python/src/wallet.rs [hunk 4]
```diff
@@ -47,7 +47,7 @@ pub fn create_wallet(options: String) -> Result<Wallet> {
 pub fn call_wallet_method(wallet: &Wallet, method: String) -> Result<String> {
     let method = serde_json::from_str::<WalletMethod>(&method)?;
     let response = crate::block_on(async {
-        match wallet.wallet.read().await.as_ref() {
+        match wallet.wallet.write().await.as_mut() {
```
Comment (Thoralf-M) [inline bindings/python/src/wallet.rs]: Mutability needed here?

**Review by Thoralf-M — CHANGES_REQUESTED** [review]
Failing test

File: bindings/core/src/method_handler/wallet.rs [hunk 5]
```diff
@@ -18,10 +18,10 @@ use crate::{method::WalletMethod, response::Response};
 /// Call a wallet method.
 pub(crate) async fn call_wallet_method_internal(wallet: &Wallet, method: WalletMethod) -> crate::Result<Response> {
     let response = match method {
-        WalletMethod::Accounts => Response::OutputsData(wallet.data().await.accounts().cloned().collect()),
+        WalletMethod::Accounts => Response::OutputsData(wallet.ledger().await.accounts().cloned().collect()),
         #[cfg(feature = "stronghold")]
         WalletMethod::Backup { destination, password } => {
-            wallet.backup(destination, password).await?;
+            wallet.backup_to_stronghold(destination, password).await?;
```
Comment (Thoralf-M) [inline bindings/core/src/method_handler/wallet.rs]: Please make a separate PR for such renaming next time

File: bindings/core/src/method_handler/wallet.rs [hunk 6]
```diff
@@ -399,8 +396,13 @@ pub(crate) async fn call_wallet_method_internal(wallet: &Wallet, method: WalletM
             let transaction = wallet.send_outputs(outputs, options).await?;
             Response::SentTransaction(TransactionWithMetadataDto::from(&transaction))
         }
+        // TODO #1934: create issue
```
Comment (Thoralf-M) [inline bindings/core/src/method_handler/wallet.rs]: TODO

File: bindings/core/src/method_handler/wallet.rs (hunk 5 above)
Comment (thibault-martinez) [reply bindings/core/src/method_handler/wallet.rs]: I would go as far as reverting it in this PR tbh. If we change the method name, we need to change the binding variant and then the bindings name.

File: bindings/core/src/method_handler/wallet.rs (hunk 6 above)
Comment (Thoralf-M) [reply bindings/core/src/method_handler/wallet.rs]: Are you actually sure that we need this?

File: cli/src/cli.rs [hunk 7]
```diff
@@ -494,10 +496,14 @@ pub async fn restore_command_stronghold(
...
     let password = get_password("Stronghold backup password", false)?;
-    if let Err(e) = wallet.restore_backup(backup_path.into(), password, None, None).await {
+    if let Err(e) = wallet
+        .restore_from_stronghold_backup(backup_path.into(), password, None, None)
+        .await
+    {
         // Clean up the file system after a failed restore (typically produces a wallet without a secret manager).
         // TODO: a better way would be to not create any files/dirs in the first place when it's not clear yet whether
         // the restore will be successful.
+        // TODO #1934: create issue for this ^
```
Comment (Thoralf-M) [inline cli/src/cli.rs]: TODO

File: bindings/core/src/method_handler/wallet.rs (hunk 5 above)
Comment (Alex6323) [reply bindings/core/src/method_handler/wallet.rs]: :shrug:

File: sdk/tests/wallet/migrate_stronghold_snapshot_v2_to_v3.rs [hunk 8]
```diff
@@ -84,7 +84,7 @@ async fn stronghold_snapshot_v2_v3_migration() {
         ]
     );
 
-    let restore_manager = Wallet::builder()
+    let mut restore_manager = Wallet::builder()
```
Comment (thibault-martinez) [inline sdk/tests/wallet/migrate_stronghold_snapshot_v2_to_v3.rs]: Remove mut?

File: sdk/src/wallet/core/mod.rs [hunk 9]
```diff
@@ -46,15 +46,22 @@ use crate::{
 /// The stateful wallet used to interact with an IOTA network.
 #[derive(Debug)]
 pub struct Wallet<S: SecretManage = SecretManager> {
+    // TODO: should we maybe group the next 3 fields into a `WalletDetails` struct?
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/mod.rs]: I personally think that's not (yet) necessary but let's see what the others think

File: sdk/src/wallet/update.rs [hunk 10]
```diff
@@ -25,12 +26,34 @@ where
     crate::wallet::Error: From<S::Error>,
     crate::client::Error: From<S::Error>,
 {
-    /// Set the alias for the wallet.
-    pub async fn set_alias(&self, alias: &str) -> crate::wallet::Result<()> {
-        let mut wallet_data = self.data_mut().await;
-        wallet_data.alias = Some(alias.to_string());
+    /// Update the wallet address with a possible new Bech32 HRP and clear the inaccessible incoming transactions.
+    pub(crate) async fn update_wallet_address_hrp(&self) -> crate::wallet::Result<()> {
```
Comment (thibault-martinez) [inline sdk/src/wallet/update.rs]: I wouldn't have `wallet` in a `Wallet` method name

File: sdk/src/wallet/storage/constants.rs [hunk 11]
```diff
@@ -1,6 +1,8 @@
 // Copyright 2021 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
+// TODO: can we remove the slight inconsistencies in this file?
```
Comment (thibault-martinez) [inline sdk/src/wallet/storage/constants.rs]: Can you expand so we can answer this question?

File: sdk/tests/wallet/bech32_hrp_validation.rs [hunk 12]
```diff
@@ -56,7 +56,7 @@ async fn bech32_hrp_prepare_output() -> Result<()> {
     let error = wallet
         .prepare_output(
             OutputParams {
-                recipient_address: wallet.address().await.to_bech32_unchecked("wronghrp"),
+                recipient_address: wallet.address().await,
```
Comment (thibault-martinez) [inline sdk/tests/wallet/bech32_hrp_validation.rs]: How come this is equivalent?

File: sdk/tests/wallet/core.rs [hunk 13]
```diff
@@ -99,11 +100,7 @@ async fn changed_bip_path() -> Result<()> {
 
     // Building the wallet with another coin type needs to return an error, because a different coin type was used in
     // the existing account
-    let mismatch_err: Result<Wallet> = Err(Error::BipPathMismatch {
-        new_bip_path: Some(Bip44::new(IOTA_COIN_TYPE)),
-        old_bip_path: Some(Bip44::new(SHIMMER_COIN_TYPE)),
-    });
-    assert!(matches!(err, mismatch_err));
+    assert!(matches!(result, Err(Error::BipPathMismatch { .. })));
```
Comment (thibault-martinez) [inline sdk/tests/wallet/core.rs]: Why make the error less precise? If that's supposed to be fixed later in the TODO, I wouldn't actually change this since it won't fail anyway. Feels like we could forget to revert to a more precise error later.

File: sdk/src/wallet/core/mod.rs (hunk 9 above)
Comment (Thoralf-M) [reply sdk/src/wallet/core/mod.rs]: I agree, as long as there are no more fields like them I would leave it as it is

File: sdk/src/wallet/operations/balance.rs [hunk 14]
```diff
@@ -27,7 +27,8 @@ where
         let protocol_parameters = self.client().get_protocol_parameters().await?;
         let slot_index = self.client().get_slot_index().await?;
 
-        let wallet_data = self.data().await.clone();
+        let wallet_address = self.address().await;
+        let wallet_ledger = self.ledger().await.clone();
```
Comment (thibault-martinez) [inline sdk/src/wallet/operations/balance.rs:31]: I know it was already there but ledger is something quite massive for a clone... Wonder if the tradeoff of keeping the lock alive is worth it, probably is since it's a read.

File: sdk/tests/wallet/bech32_hrp_validation.rs (hunk 12 above)
Comment (Alex6323) [reply sdk/tests/wallet/bech32_hrp_validation.rs]: it's not

File: sdk/src/wallet/storage/constants.rs (hunk 11 above)
Comment (Alex6323) [reply sdk/src/wallet/storage/constants.rs]: just the names seemed a bit inconsistent. `default_storage_path` is either returning something that is called a folder name or called a path. Should return relative paths in any case I think.

File: sdk/src/wallet/operations/balance.rs (hunk 14 above)
Comment (Alex6323) [reply sdk/src/wallet/operations/balance.rs:31]: I agree. I think cloning isn't worth it tbh, but we probably did it for a good reason ... like holding locks across .await points and such. It can't write to the ledger as long as there's a read lock, so this could potentialy dead lock (again?)

File: sdk/src/wallet/update.rs (hunk 10 above)
Comment (Alex6323) [reply sdk/src/wallet/update.rs]: okay

File: bindings/core/src/method_handler/wallet.rs (hunk 6 above)
Comment (Alex6323) [reply bindings/core/src/method_handler/wallet.rs]: No I don't think there's a definitive reason for adding this. I removed this commented code. We could still add an issue if you think it was a valid idea.

File: sdk/src/wallet/operations/balance.rs (hunk 14 above)
Comment (Thoralf-M) [reply sdk/src/wallet/operations/balance.rs:31]: Yes was done as part of the changes to reduce the possibility of deadlocks

File: bindings/core/src/method_handler/wallet.rs (hunk 6 above)
Comment (Thoralf-M) [reply bindings/core/src/method_handler/wallet.rs]: Currently I don't see when one would need to set it later, so personally I wouldn't open an issue

File: sdk/tests/wallet/core.rs [hunk 15]
```diff
@@ -97,13 +97,14 @@ async fn changed_bip_path() -> Result<()> {
         .finish()
         .await;
 
-    // Building the wallet with another coin type needs to return an error, because a different coin type was used in
-    // the existing account
-    let mismatch_err: Result<Wallet> = Err(Error::BipPathMismatch {
+    let _mismatch_err: Result<()> = Err(Error::BipPathMismatch {
```
Comment (thibault-martinez) [inline sdk/tests/wallet/core.rs]: any reason for the underscore?

Comment (Alex6323) [reply sdk/tests/wallet/core.rs]: clippy warning (bc the binding is not really used, just matched)

File: sdk/src/wallet/core/operations/address_generation.rs [hunk 16]
```diff
@@ -25,7 +25,7 @@ impl Wallet {
         address_index: u32,
         options: impl Into<Option<GenerateAddressOptions>> + Send,
     ) -> crate::wallet::Result<Ed25519Address> {
-        // TODO #1279: not sure yet whether we also should allow this method to generate addresses for different bip
+        // TODO: not sure yet whether we also should allow this method to generate addresses for different bip
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/operations/address_generation.rs:28]: Is this not linked to an issue anymore?

Comment (Thoralf-M) [reply sdk/src/wallet/core/operations/address_generation.rs:28]: Not sure why the TODO is modified here, but the whole file is deleted in https://github.com/iotaledger/iota-sdk/pull/1586

File: sdk/tests/wallet/core.rs (hunk 15 above)
Comment (thibault-martinez) [reply sdk/tests/wallet/core.rs]: That sucks, sounds like a clippy issue tbh, if ever you want to report it

File: sdk/src/wallet/core/operations/address_generation.rs (hunk 16 above)
Comment (thibault-martinez) [reply sdk/src/wallet/core/operations/address_generation.rs:28]: Ah, right, let's ignore then

Comment (Alex6323) [reply sdk/src/wallet/core/operations/address_generation.rs:28]: no, it isn't ... it was something that I thought I could figure out during the "single account wallet" PR. It's not relevant anymore here as this file gets removed with https://github.com/iotaledger/iota-sdk/pull/1586. For generating addresses on the secret manager there's no reason to debate this.

File: sdk/tests/wallet/core.rs (hunk 15 above)
Comment (Alex6323) [reply sdk/tests/wallet/core.rs]: it definitely struck we as a bit weird, but I guess they are technically right.

File: sdk/src/wallet/core/mod.rs [hunk 17]
```diff
@@ -360,14 +345,22 @@ where
     crate::wallet::Error: From<S::Error>,
     crate::client::Error: From<S::Error>,
 {
+    // TODO: Is this method needed? It calls `get_default_sync_options` which the builder does not.
```
Comment (Thoralf-M) [inline sdk/src/wallet/core/mod.rs]: Looks like this got messed up in the single account merge since it was an account
`get_default_sync_options()` should be loaded in the wallet builder now

File: sdk/src/wallet/core/mod.rs (hunk 9 above)
Comment (Alex6323) [reply sdk/src/wallet/core/mod.rs]: removed the comment

File: sdk/src/wallet/storage/constants.rs (hunk 11 above)
Comment (Alex6323) [reply sdk/src/wallet/storage/constants.rs]: I can remove the TODO tbh

File: sdk/src/wallet/operations/balance.rs (hunk 14 above)
Comment (Alex6323) [reply sdk/src/wallet/operations/balance.rs:31]: yeah, I kept it

File: sdk/src/wallet/core/mod.rs (hunk 17 above)
Comment (Thoralf-M) [reply sdk/src/wallet/core/mod.rs]: Opened an issue for it https://github.com/iotaledger/iota-sdk/issues/2022

Comment (Thoralf-M) [reply sdk/src/wallet/core/mod.rs]: So this method isn't needed

Comment (Alex6323) [reply sdk/src/wallet/core/mod.rs]: and there it goes ....

File: sdk/src/wallet/core/operations/stronghold_backup/mod.rs [hunk 18]
```diff
@@ -64,7 +66,7 @@ impl Wallet {
     /// coin type doesn't match
     /// If a bech32 hrp is provided to ignore_if_bech32_hrp_mismatch, that doesn't match the one of the current address,
     /// the wallet will not be restored.
-    pub async fn restore_backup(
+    pub async fn restore_from_stronghold_backup(
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/operations/stronghold_backup/mod.rs]: Also revert?

File: sdk/tests/wallet/core.rs (hunk 15 above)
Comment (thibault-martinez) [reply sdk/tests/wallet/core.rs]: I don't think they are tbh, it's clearly being used 🤷🏻‍♂️

File: sdk/src/wallet/operations/balance.rs (hunk 14 above)
Comment (thibault-martinez) [reply sdk/src/wallet/operations/balance.rs:31]: I will open an issue then, we should definitely find a way to not clone this

Comment (thibault-martinez) [reply sdk/src/wallet/operations/balance.rs:31]: https://github.com/iotaledger/iota-sdk/issues/2024

File: sdk/src/wallet/storage/constants.rs (hunk 11 above)
Comment (thibault-martinez) [reply sdk/src/wallet/storage/constants.rs]: As you want, either remove it, address it or create an issue

File: sdk/src/wallet/operations/balance.rs [hunk 19]
```diff
@@ -36,20 +37,21 @@ where
...
 
-        let claimable_outputs = wallet_data.claimable_outputs(OutputsToClaim::All, slot_index, &protocol_parameters)?;
+        let claimable_outputs =
+            wallet_ledger.claimable_outputs(&wallet_address, OutputsToClaim::All, slot_index, &protocol_parameters)?;
 
         #[cfg(feature = "participation")]
         {
             if let Some(voting_output) = &voting_output {
-                if voting_output.output.as_basic().address() == wallet_data.address.inner() {
+                if voting_output.output.as_basic().address() == self.address.read().await.inner() {
```
Comment (thibault-martinez) [inline sdk/src/wallet/operations/balance.rs]: Reuse `wallet_address` ?

File: sdk/src/wallet/update.rs [hunk 20]
```diff
@@ -25,12 +26,34 @@ where
...
+
+            log::debug!("[save] wallet ledger with updated bech32 hrp",);
+            self.storage_manager().save_wallet_ledger(&wallet_ledger).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Update the alias for the wallet.
+    pub async fn update_wallet_alias(&self, alias: &str) -> crate::wallet::Result<()> {
```
Comment (thibault-martinez) [inline sdk/src/wallet/update.rs]: `update_alias`

Comment (thibault-martinez) [reply sdk/src/wallet/update.rs]: I know you probably changed set to update to match `update_address_hrp` but I personally prefer set. I'd rename everything to set.

File: sdk/src/wallet/core/mod.rs [hunk 21]
```diff
@@ -46,15 +46,21 @@ use crate::{
 /// The stateful wallet used to interact with an IOTA network.
 #[derive(Debug)]
 pub struct Wallet<S: SecretManage = SecretManager> {
+    pub(crate) address: Arc<RwLock<Bech32Address>>,
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/mod.rs:49]: Do we even want these 3 to be mutable?

Comment (Thoralf-M) [reply sdk/src/wallet/core/mod.rs:49]: address and bip_path I would say not really, but with the current design we need it when we restore backups, we might can change it when the backup creates a new wallet instead of modifying the current one. Similar discussion here to https://github.com/iotaledger/iota-sdk/pull/1756#discussion_r1476054598

Comment (thibault-martinez) [reply sdk/src/wallet/core/mod.rs:49]: Do we have an issue to remove the mutability at a later point then?

File: sdk/src/wallet/core/mod.rs [hunk 22]
```diff
@@ -410,32 +369,57 @@ where
...
-        self.data.read().await
+    /// Get the wallet address config.
+    pub async fn address(&self) -> Bech32Address {
+        self.address.read().await.clone()
     }
 
-    pub(crate) async fn data_mut(&self) -> tokio::sync::RwLockWriteGuard<'_, WalletData> {
-        self.data.write().await
+    /// Get the wallet address config mutably.
+    pub async fn address_mut(&self) -> tokio::sync::RwLockWriteGuard<'_, Bech32Address> {
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/mod.rs]: I would remove this, shouldn't be part of the public API

File: sdk/src/wallet/core/mod.rs [hunk 23]
```diff
@@ -410,32 +369,57 @@ where
...
+        self.address.read().await.hrp
+    }
+
+    /// Get the wallet's configured bip path.
+    pub async fn bip_path(&self) -> Option<Bip44> {
+        *self.bip_path.read().await
+    }
+
+    /// Get the wallet's configured bip path mutably.
+    pub async fn bip_path_mut(&self) -> tokio::sync::RwLockWriteGuard<'_, Option<Bip44>> {
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/mod.rs]: I would remove this, shouldn't be part of the public API

File: sdk/src/wallet/core/operations/stronghold_backup/stronghold_snapshot.rs [hunk 24]
```diff
@@ -1,73 +1,110 @@
...
+pub(crate) const WALLET_ALIAS_KEY: &str = "wallet_alias";
 
-impl<S: 'static + SecretManagerConfig> Wallet<S> {
-    pub(crate) async fn store_data_to_stronghold(&self, stronghold: &StrongholdAdapter) -> crate::wallet::Result<()> {
+impl<S: 'static + SecretManagerConfig> Wallet<S>
+where
+    crate::wallet::Error: From<S::Error>,
+    crate::client::Error: From<S::Error>,
+{
+    pub(crate) async fn backup_wallet_to_stronghold_snapshot(
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/operations/stronghold_backup/stronghold_snapshot.rs]: Either revert or at least remove wallet from the name

File: sdk/src/wallet/storage/manager.rs [hunk 25]
```diff
@@ -122,35 +127,46 @@ mod tests {
...
         let storage_manager = StorageManager::new(Memory::default(), None).await.unwrap();
-        assert!(storage_manager.load_wallet_data().await.unwrap().is_none());
+        assert!(storage_manager.load_wallet_ledger().await.unwrap().is_none());
 
-        let wallet_data = WalletData::mock();
+        let wallet_ledger = WalletLedger::test_instance();
 
-        storage_manager.save_wallet_data(&wallet_data).await.unwrap();
-        let wallet = storage_manager.load_wallet_data().await.unwrap();
-        assert!(matches!(wallet, Some(data) if data.alias == Some("Alice".to_string())));
```
Comment (thibault-martinez) [inline sdk/src/wallet/storage/manager.rs:133]: We lost the alias check?

Comment (thibault-martinez) [reply sdk/src/wallet/storage/manager.rs:133]: Ah it got moved, right

File: sdk/src/wallet/core/mod.rs (hunk 21 above)
Comment (Thoralf-M) [reply sdk/src/wallet/core/mod.rs:49]: No, it's also not clear if we want to change it this way (at least to me)

File: sdk/src/wallet/core/operations/stronghold_backup/mod.rs [hunk 26]
```diff
@@ -79,42 +81,40 @@ impl Wallet {
...
-
-        let read_bip_path = read_wallet_data.as_ref().and_then(|data| data.bip_path);
+        let (read_address, read_bip_path, read_alias, read_client_options, read_secret_manager, read_wallet_ledger) =
+            restore_from_stronghold_snapshot::<SecretManager>(&new_stronghold).await?;
 
         // If the bip path is not matching the current one, we may ignore the backup
         let ignore_backup_values = ignore_if_bip_path_mismatch.map_or(false, |ignore| {
             if ignore {
-                // TODO: #1279 okay that if both are none we always load the backup values?
+                // TODO: is it okay that if both are none we always load the backup values?
```
Comment (Thoralf-M) [inline sdk/src/wallet/core/operations/stronghold_backup/mod.rs:109]: Do we have an issue for this?

File: sdk/src/wallet/core/operations/stronghold_backup/stronghold_snapshot.rs [hunk 27]
```diff
@@ -1,73 +1,110 @@
...
-        stronghold.set(WALLET_DATA_KEY, &serialized_wallet_data).await?;
+        // Store the wallet address
+        stronghold
+            .set(WALLET_ADDRESS_KEY, self.address().await.as_ref())
+            .await?;
+
+        // Store the wallet bip path
+        stronghold.set(WALLET_BIP_PATH_KEY, &self.bip_path().await).await?;
+
+        // Store the wallet alias)
```
Comment (Thoralf-M) [inline sdk/src/wallet/core/operations/stronghold_backup/stronghold_snapshot.rs]:
```suggestion
        // Store the wallet alias
```

File: sdk/src/wallet/update.rs [hunk 28]
```diff
@@ -25,12 +26,34 @@ where
...
+        log::debug!("updating wallet with new bech32 hrp: {}", bech32_hrp);
+
+        self.address_mut().await.hrp = bech32_hrp;
+
         #[cfg(feature = "storage")]
-        self.storage_manager().save_wallet_data(&wallet_data).await?;
+        {
+            let wallet_ledger = {
+                let mut wallet_ledger = self.ledger_mut().await;
+                wallet_ledger.inaccessible_incoming_transactions.clear();
```
Comment (thibault-martinez) [inline sdk/src/wallet/update.rs]: This was previously done outside of the storage feature?

File: sdk/src/wallet/update.rs (hunk 20 above)
Comment (Alex6323) [reply sdk/src/wallet/update.rs]: this is done

File: sdk/src/wallet/core/mod.rs (hunk 21 above)
Comment (Alex6323) [reply sdk/src/wallet/core/mod.rs:49]: created an issue: https://github.com/iotaledger/iota-sdk/issues/2031

File: sdk/src/wallet/core/mod.rs [hunk 29]
```diff
@@ -360,37 +344,12 @@ where
     crate::wallet::Error: From<S::Error>,
     crate::client::Error: From<S::Error>,
 {
-    /// Create a new wallet.
-    pub(crate) async fn new(inner: Arc<WalletInner<S>>, data: WalletData) -> Result<Self> {
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/mod.rs:364]: Why wasn't it used anymore btw?

File: sdk/src/wallet/core/mod.rs [hunk 30]
```diff
@@ -410,32 +369,55 @@ where
         self.inner.emit(wallet_event).await
     }
 
-    pub async fn data(&self) -> tokio::sync::RwLockReadGuard<'_, WalletData> {
-        self.data.read().await
+    /// Get the wallet address config.
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/mod.rs]: config?

File: sdk/src/wallet/core/mod.rs [hunk 31]
```diff
@@ -410,32 +369,55 @@ where
...
-    pub(crate) async fn data_mut(&self) -> tokio::sync::RwLockWriteGuard<'_, WalletData> {
-        self.data.write().await
+    pub(crate) async fn address_mut(&self) -> tokio::sync::RwLockWriteGuard<'_, Bech32Address> {
+        self.address.write().await
     }
 
-    #[cfg(feature = "storage")]
-    pub(crate) fn storage_manager(&self) -> &StorageManager {
-        &self.storage_manager
+    /// Get the wallet's configured Bech32 HRP.
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/mod.rs]:
```suggestion
    /// Get the wallet's Bech32 HRP.
```

File: sdk/src/wallet/core/mod.rs [hunk 32]
```diff
@@ -410,32 +369,55 @@ where
...
+    pub async fn bech32_hrp(&self) -> Hrp {
+        self.address.read().await.hrp
+    }
+
+    /// Get the wallet's configured bip path.
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/mod.rs]:
```suggestion
    /// Get the wallet's bip path.
```

File: sdk/src/wallet/core/mod.rs (hunk 21 above)
Comment (Alex6323) [reply sdk/src/wallet/core/mod.rs:49]: created an issue: https://github.com/iotaledger/iota-sdk/issues/2032

File: sdk/src/wallet/core/operations/stronghold_backup/mod.rs (hunk 26 above)
Comment (Alex6323) [reply sdk/src/wallet/core/operations/stronghold_backup/mod.rs:109]: created an issue: https://github.com/iotaledger/iota-sdk/issues/2032

File: sdk/src/wallet/core/mod.rs (hunk 29 above)
Comment (Thoralf-M) [reply sdk/src/wallet/core/mod.rs:364]: Wasn't moved when doing the single account change https://github.com/iotaledger/iota-sdk/issues/2022

File: sdk/src/wallet/core/mod.rs (hunk 30 above)
Comment (Alex6323) [reply sdk/src/wallet/core/mod.rs]: remnant

File: sdk/src/wallet/update.rs (hunk 28 above)
Comment (Alex6323) [reply sdk/src/wallet/update.rs]: good catch

Comment (Alex6323) [reply sdk/src/wallet/update.rs]: fixed

File: sdk/src/wallet/core/mod.rs (hunk 30 above)
Comment (Alex6323) [reply sdk/src/wallet/core/mod.rs]: removed

File: sdk/src/wallet/core/mod.rs (hunk 32 above)
Comment (Alex6323) [reply sdk/src/wallet/core/mod.rs]: Seems like it was committed but not auto-closed.

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/wallet/core/mod.rs (hunk 32 above)
Comment (thibault-martinez) [reply sdk/src/wallet/core/mod.rs]: It was not committed

Comment (Alex6323) [reply sdk/src/wallet/core/mod.rs]: ooh, I see

File: sdk/src/wallet/core/operations/stronghold_backup/stronghold_snapshot.rs [hunk 33]
```diff
@@ -1,73 +1,110 @@
...
+
+        let serialized_wallet_ledger = serde_json::to_value(&WalletLedgerDto::from(&*self.ledger.read().await))?;
+        stronghold.set(WALLET_LEDGER_KEY, &serialized_wallet_ledger).await?;
 
         Ok(())
     }
 }
 
-pub(crate) async fn read_wallet_data_from_stronghold_snapshot<S: 'static + SecretManagerConfig>(
+pub(crate) async fn restore_from_stronghold_snapshot<S: 'static + SecretManagerConfig>(
```
Comment (thibault-martinez) [inline sdk/src/wallet/core/operations/stronghold_backup/stronghold_snapshot.rs:65]: I am a bit annoying now but this one is not a Wallet method so wallet in the name kinda makes sense 😬 
Otherwise we don't really know what we are restoring

Comment (thibault-martinez) [reply sdk/src/wallet/core/operations/stronghold_backup/stronghold_snapshot.rs:65]: Meh, it's not public anyway, let's ignore that's fine

**Review by Thoralf-M — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

File: sdk/src/wallet/core/operations/stronghold_backup/stronghold_snapshot.rs (hunk 33 above)
Comment (Alex6323) [reply sdk/src/wallet/core/operations/stronghold_backup/stronghold_snapshot.rs:65]: can be changed back in #1941
